lib/ecto/query.ex

```diff
@@ -980,6 +992,69 @@ defmodule Ecto.Query do
     [on: expr]
   end
 
+  @doc """
+  A recursive common table expression (CTE) also known as WITH expression.
+
+  For keyword queries use `:with` or `:with_recursive` keywords.
+
+  ## Options
+
+  * `:as` – CTE name. Required.
+  * `:recursive` – boolean indicating whether WITH expression is recursive.
+
+  ## Keyword examples
+
+  Category tree where each subcategory references its parent by `parent_id` with products count:
+
+    category_tree_query =
+      from c in Category,
+        where: is_nil(c.parent_id),
+        order_by: c.name,
+        select: %{c | depth: fragment("1")},
+        union_all: (
+          from c in Category,
+            join: r in "category_tree",
+            where: r.id == c.parent_id,
+            order_by: c.name,
+            select: %{c | depth: fragment("? + 1", r.depth)}
+        )
+      )
+
+    from c in "category_tree",
+      with_recursive: category_tree_query,
+      as: "category_tree",
+      left_join: p in assoc(c, :products),
+      group_by: c.id,
+      select: %{c | products_count: count(p.id)}
+
+  ## Expression examples
+
+  Products and their category names for breadcrumbs:
+
+    Product
+    |> cte(category_tree_query, as: "category_tree", recursive: true)
+    |> join(:left, [p], c in "category_tree", on: c.id == p.category_id)
+    |> group_by([p], p.id)
+    |> select([p, c], %{p | category_names: fragment("ARRAY_AGG(?)", c.name)})
+  """
+  def cte(%__MODULE__{with: with_expr} = query, with_query, opts) do
+    unless opts[:as], do: raise "`:as` option is required in `cte/3`"
+
+    with_expr = with_expr || %WithExpr{}
+
+    with_expr = %{
+      with_expr
+      | queries: with_expr.queries ++ [%WithQueryExpr{query: with_query, as: opts[:as]}],
```

Comment:
I would like to store this as `[{key, query}]`. This way we don't need the `WithQueryExpr` struct. Mirroring the structure we see in windows.